src/main/java/uk/gov/justice/digital/hmpps/cmd/api/service/NotificationService.kt

```diff
@@ -22,29 +23,55 @@ class NotificationService(@Autowired val shiftNotificationRepository: ShiftNotif
 
-        val now = LocalDate.now(clock)
-        val from = fromParam.orElse(now.withDayOfMonth(1)).atTime(LocalTime.MIN)
-        val to = toParam.orElse(now.plusMonths(plusMonths).withDayOfMonth(now.lengthOfMonth())).atTime(LocalTime.MAX)
-        log.debug("Finding UnreadOnly: ($unreadOnly) Notifications between ($from) and ($to)")
-
-        val shifts = shiftNotificationRepository.findAllByQuantumIdAndLastModifiedDateTimeIsBetween(quantumId, from, to)
-        val details = shiftTaskNotificationRepository.findAllByQuantumIdAndLastModifiedDateTimeIsBetween(quantumId, from, to)
-        log.debug("Found (${shifts.size}) Shift and (${details.size}) Detail Notifications")
+        val shiftDtos = getShiftNotifications(quantumId, fromDateTime, toDateTime, unreadOnly)
+        val taskDtos = getShiftTaskNotifications(quantumId, fromDateTime, toDateTime, unreadOnly)
+        log.debug("Found (${shiftDtos.size}) Shift and (${taskDtos.size}) Task Notifications")
+        return shiftDtos.union(taskDtos)
+    }
 
-        val shiftDtos = NotificationDto.fromShift(shifts.filter { !unreadOnly || (unreadOnly && !it.read) })
-        val detailDtos = NotificationDto.fromTask(details.filter { !unreadOnly || (unreadOnly && !it.read) })
+    private fun getShiftTaskNotifications(quantumId: String, fromDateTime: LocalDateTime, toDateTime: LocalDateTime, unreadOnly: Boolean): List<NotificationDto> {
+        val tasks = shiftTaskNotificationRepository.findAllByQuantumIdAndLastModifiedDateTimeIsBetween(
+                quantumId,
+                fromDateTime,
+                toDateTime).filter { filterUnread(unreadOnly, it.read) }
+        val taskDtos = NotificationDto.fromTasks(tasks)
+        tasks.forEach { it.read = true }
+        shiftTaskNotificationRepository.saveAll(tasks)
+        return taskDtos
+    }
 
+    private fun getShiftNotifications(quantumId: String, fromDateTime: LocalDateTime, toDateTime: LocalDateTime, unreadOnly: Boolean): List<NotificationDto> {
+        val shifts = shiftNotificationRepository.findAllByQuantumIdAndLastModifiedDateTimeIsBetween(
+                quantumId,
+                fromDateTime,
+                toDateTime).filter { filterUnread(unreadOnly, it.read) }
+        val shiftDtos = NotificationDto.fromShifts(shifts)
         shifts.forEach { it.read = true }
-        details.forEach { it.read = true }
-
         shiftNotificationRepository.saveAll(shifts)
-        shiftTaskNotificationRepository.saveAll(details)
-        log.debug("Updated (${shifts.size}) Shift and (${details.size}) Detail Notifications")
+        return shiftDtos
+    }
 
-        return shiftDtos.union(detailDtos)
+    private fun calculateFromAndToDates(fromParam: Optional<LocalDate>, toParam: Optional<LocalDate>): Pair<LocalDateTime, LocalDateTime> {
+        val now = LocalDate.now(clock)
+
+        // Default 'from' date is day 1 of the current month
+        val from = fromParam.orElse(now.withDayOfMonth(1))
```

Comment:
If the from date is not supplied, and the two date is supplied, do we want the from date to be set to 3 months before the to date

eg
```
from = null
to = 2020-05-29

// returns
from = 2020-02-30
to = 2020-05-29
```

Reply:
Actually I think we need our default to date to be relative otherwise we can end up with a from date thats after our to date

```
from = null
to = 1999-01-01

// returns
from = 2020-07-01
to = 1999-01-01
```

Reply:
I have added better to and from handling and written tests to cover the scenarios